nada_mir/proto/nillion/nada/v1/operations.proto

```diff
@@ -0,0 +1,216 @@
+  REVEAL = 1;
+  // Not operation variant
+  NOT = 2;
+}
+
+// MIR Binary operation.
+// Basically most arithmetic operations: Addition, Subtraction
+// Division, Modulo, Power, etc.
+message BinaryOperation {
+  // Operation variant
+  BinaryOperationVariant variant = 1;
+  // Left operand of the operation
+  uint64 left = 2;
+  // Right operand of the operation
+  uint64 right = 3;
+}
+
+// Represents a MIR Unary operation:
+// - Cast
+// - Not
+// - Reveal
+// - Unzip
+message UnaryOperation {
+  // Operation variant
+  UnaryOperationVariant variant = 1;
+  // The operand of the operation
+  uint64 this = 2;
+}
+
+message IfElseOperation {
+  // operand of the conditional operation
+  uint64 cond = 1;
+  // operand of the first operation
+  uint64 first = 2;
+  // operand of the second operation
+  uint64 second = 3;
+}
+
+message RandomOperation {}
+
+// Input reference structure, can be used for:
+// - Input
+// - Literal
+// Also, it is used to describe the nada function arguments.
+message InputReference {
+  // Index of the input/literal operation referred by this operation
+  string refers_to = 1;
+}
+
+
+message LiteralReference {
+  // Literal refers to
+  string refers_to = 1;
+}
+
+message MapOperation {
+  // Function to execute
+  uint64 fn = 1;
+  // Map operation child
+  uint64 child = 2;
+}
+
+message ReduceOperation {
+  // Function to execute
+  uint64 fn = 1;
+  // Reduce operation child
+  uint64 child = 2;
+  // Initial accumulator value
+  uint64 initial = 3;
+}
+
+message NewOperation {
+  // The elements of this compound type
+  repeated uint64 elements = 1;
+}
+
+message ArrayAccessor {
+  // array index - for now an integer but eventually it could be the result of
+  // an operation
+  uint32 index = 1;
+  // source - The Operation that represents the array we are accessing
+  uint64 source = 2;
+}
+
+enum TupleIndex {
+  // The left element of the tuple
+  LEFT = 0;
+  // The right element of the tuple
+  RIGHT = 1;
+}
+
+message TupleAccessor {
+  // tuple index (left or right)
+  TupleIndex index = 1;
+  // source - The Operation that represents the tuple we are accessing
+  uint64 source = 2;
+}
+
+message NtupleAccessor {
+  // tuple index (left or right)
+  uint32 index = 1;
+  // source - The Operation that represents the tuple we are accessing
+  uint64 source = 2;
+}
+
+message ObjectAccessor {
+  // object key
+  string key = 1;
+  // source - The Operation that represents the object we are accessing
+  uint64 source = 2;
+}
+
+message NadaFunctionArgRef {
+  // Function owner of this argument
+  uint64 function_id = 1;
+  // Refers to the argument
+  string refers_to = 2;
+}
+
+message CastOperation {
+  // The operand of the operation
+  uint64 target = 1;
+  // The type to cast to
+  nillion.nada.types.v1.NadaType cast_to = 2;
+}
+
+// The Operation.
+// An operation is identified by:
+// - The operation variant
+message Operation {
+    // Operation identifier
+    uint64 id = 1;
```

Comment:
Is this the same identifier as the one in `ProgramMir.operations`? If so it seems redundant and can save at least 2 bytes per operation.

Reply:
Yes it is, I deleted it.